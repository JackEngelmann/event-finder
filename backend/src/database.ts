import sqlite3 from 'sqlite3'

sqlite3.verbose()

const clubNames = [
'Kathys Garage',
'Stolperdiele',
'LOBO',
'Kraftwerk Mitte',
'Downtown',
'Rosis',
'Altes WettbÃ¼ro',
'Standesamt ',
'Gutzkow',
'Club Mensa',
'RÃ¶sslstube (?)',
'Sektor',
'Objekt klein a',
'Blue Dance Club',
'Gisela ',
'Arteum',
'Aquarium',
'BÃ¤renzwinger (?)',
'My House',
'TBA',
'Puro ',
'Club Neu',
'Groovestation',
'Wu5 ',
'Novitatis',
'COunt down ',
'Club 11',
'TraumtÃ¤nzer',
'GAG 18',
'Borsi 34',
'HÃ¤nge Mathe',
'Scheune Dresden',
'Parkhotel Dresden',
'Paula',
'Showboxx',
'Eventwerk (?)',
'Zille',
'Chemiefabrik',
'Reithalle StraÃŸe E',
'Alter Schlachthof ',
'Puschkin',
'Tane Ju',
'Bunker (?) ',
'Beatpol',
'Ostpol',
'Der Lude',
]

const DB_SOURCE = 'db.sqlite'
const database = new sqlite3.Database(DB_SOURCE, err => {
    if (err) {
        console.error(err.message)
        throw err
    }
    console.log('Connected to SQLite database')
    initializeClubTable()
    initializeEventTable()
})

function initializeClubTable() {
    console.log('initializeClubTable')
    const sql = `
        CREATE TABLE club (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name text
        )
    `
    database.run(sql, err => {
        if (err) return // table was created before
        const sql = 'INSERT INTO club (name) VALUES (?)'
        clubNames.forEach(clubName => {
            database.run(sql, [clubName])
        })
    })
}

function initializeEventTable() {
    console.log('initializeEventTable')
    const sql = `
        CREATE TABLE event (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            name text,
            description text,
            date string,
            clubId integer,
            imageUrl text
        )
    `
    database.run(sql, err => {
        if (err) return // table was created before
        const sql = 'INSERT INTO event (name, description, date, clubId, imageUrl) VALUES (?, ?, ?, ?, ?)'
        database.run(sql, [
            'Highway to Hell - Halloween 2019',
`â˜  HIGHWAY TO HELL â˜ 
Traditionell am 30.10. erwarten wir Euch wieder zu unserer spektakulÃ¤ren Halloween Party Part II  :
ðŸŽƒ SPECIAL HALLOWEEN DEKO ðŸŽƒ GRUSEL TUNNEL  
ðŸŽƒANIMATION ðŸŽƒ MAKE UP MAKER 
ðŸŽƒSPECIAL HALLOWEEN DRINK 

 â˜  â˜  â˜  LINE UP & MUSIC - 2 FLOORS â˜  â˜  â˜  
Hip Hop, Charts, 90s, Electro,  Party Hits, Latin
#Louis_FOX . Red A . DJ T-vine
                  ðŸš¨ðŸš¨EINTRITTðŸš¨ðŸš¨ 
Eintritt frei bis 00:00 Uhr, danach mit KostÃ¼m 3â‚¬ ohne 5â‚¬`,
            '2019-10-24T23:00:00.000Z',
            1,
            'images/eventLobo.jpg'
        ])
        database.run(sql, [
            'ROUTINEÃ—Shape',
            `Wer dieser Tage in Form kommen will, muss Format haben. Doch der umgebende Raum ist es, in den es sich auszudehnen lohnt, so er es zulÃ¤sst - und umso mehr, wenn nicht. In vollem Umfang den Rahmen zu sprengen bedeutet, das Ganze vom Rand her zu denken. Was zunÃ¤chst wie der Umriss einer HÃ¼lle erscheint, entpuppt sich bei genauerer Betrachtung als das Wesen der Gestalt. Und wo schlieÃŸlich alles aus dem Leim geht, wird die Dimension, wird das ganze AusmaÃŸ der seelischen Katastrophe sichtbar. So erst bekommt die Sache Kontur. Routinen kÃ¶nnen hier helfen.`,
            '2019-10-24T23:00:00.000Z',
            2,
            'images/eventOka.jpg'
        ])
        database.run(sql, [
            'Rosis Teleparty - Girls Power',
`Rosis Music Television Sunday bringt die besten Musikvideos und grÃ¶ÃŸten Hits zurÃ¼ck auf die Leinwand.

Diesmal im Spotlight, die Girls der Pop & Rockmusik umrandet von weiteren groÃŸartigen Musikvideos weiterer KÃ¼nstler

EINTRITT FREI`,
            '2019-10-24T23:00:00.000Z',
            3,
            'images/eventRosis.jpg'
        ])
    })
}

export default database