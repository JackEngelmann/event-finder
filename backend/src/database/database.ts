import pg, { Client } from 'pg'
import { databaseConfig, DatabaseConfig } from '../../databaseConfig'
import { applyDbScripts } from './applyDbScripts'

export interface Database {
    get<Row = any>(text: string, values?: any): Promise<Row | undefined>
    all<Row = any>(text: string, values?: any): Promise<Row[]>
    run(text: string, values?: any): Promise<void>
}

class PostgresDatabase implements Database {
    db: pg.Client
    constructor(
        connectionString: string,
        callback?: (database: Database) => void
    ) {
        this.db = new Client({
            connectionString,
        })
        this.db.connect().then(() => {
            if (callback) callback(this)
        })
    }

    async get<Row>(text: string, values = []) {
        return this.db
            .query<Row>(text, values)
            .then(res => res.rows[0])
            .catch(err => {
                logDatabaseError(err, text)
                return err
            })
    }

    async all<Row>(text: string, values = []) {
        return this.db
            .query<Row>(text, values)
            .then(res => res.rows)
            .catch(err => {
                logDatabaseError(err, text)
                return err
            })
    }

    async run(text: string, values?: any) {
        return this.db
            .query(text, values)
            .then(res => undefined)
            .catch(err => {
                logDatabaseError(err, text)
                return err
            })
    }
}

export const createDatabase = () =>
    new PostgresDatabase(databaseConfig.connectionString, async db => {
        console.log('appydbscripts')
        await applyDbScripts(db, databaseConfig)
    })

let testDb: PostgresDatabase | undefined = undefined
export const createTestDb = async (testDbName: string) =>
    new Promise<Database>(async resolve => {
        if (!databaseConfig.getTestConnectionString) {
            throw new Error(
                'to create a test db the getTestConnectionString must be defined in the database configuration'
            )
        }
        try {
            const client = new Client({
                connectionString: databaseConfig.connectionString,
            })
            await client.connect()
            await client.query(`DROP DATABASE IF EXISTS ${testDbName}`)
            await client.query(`CREATE DATABASE ${testDbName}`)
            await client.end()
            const db = new PostgresDatabase(
                databaseConfig.getTestConnectionString(testDbName),
                async () => {
                    await applyDbScripts(db, databaseConfig)
                    testDb = db
                    resolve(db)
                }
            )
        } catch (err) {
            console.error(err)
        }
    })

export const destroyTestDb = (testDbName: string) =>
    new Promise(async resolve => {
        if (testDb) await testDb.db.end()
        const client = new Client({
            connectionString: databaseConfig.connectionString,
        })
        await client.connect()
        await client.query(`DROP DATABASE IF EXISTS ${testDbName}`)
        await client.end()
        resolve()
    })

function logDatabaseError(errorMessage: string, sqlText: string) {
    console.error('---')
    console.error(sqlText)
    console.error('leads to this error:   ------------------')
    console.error(errorMessage)
    console.error('---')
}
