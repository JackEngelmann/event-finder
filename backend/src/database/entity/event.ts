import { Connection, Entity, Column, PrimaryGeneratedColumn } from 'typeorm'

@Entity('event')
export class EventDataModel {
    @Column({ type: 'decimal', nullable: true })
    admissionFee?: number | null

    @Column({ type: 'decimal', nullable: true })
    admissionFeeWithDiscount?: number | null

    @Column({ type: 'int', nullable: true })
    amountOfFloors?: number | null

    @Column()
    clubId!: number

    @Column()
    date!: string

    @Column({
        type: 'text',
        nullable: true,
        charset: 'utf8mb4',
        collation: 'utf8mb4_general_ci',
    })
    description?: string | null

    @PrimaryGeneratedColumn()
    id!: number

    @Column({ type: 'text', nullable: true })
    imageUrl?: string | null

    @Column({ type: 'text', nullable: true })
    link?: string | null

    @Column({ type: 'int', nullable: true })
    minimumAge?: number | null

    @Column()
    name!: string

    @Column({ type: 'int', nullable: true })
    priceCategory?: number | null

    @Column({
        type: 'text',
        nullable: true,
        charset: 'utf8mb4',
        collation: 'utf8mb4_general_ci',
    })
    special?: string | null
}

export class EventModel {
    private connection: Connection

    constructor(connection: Connection) {
        this.connection = connection
    }

    async createEvent(input: {
        admissionFee?: number
        admissionFeeWithDiscount?: number
        amountOfFloors?: number
        clubId: number
        date: string
        description?: String
        genreIds?: number[]
        imageUrl?: string
        link?: string
        minimumAge?: number
        name: string
        priceCategory?: number
        special?: string
    }) {
        const event = new EventDataModel()
        Object.assign(event, input)
        await this.connection.manager.save(event)

        return event.id
    }

    async deleteEvent(id: number) {
        const event = await this.connection.manager.findOneOrFail(
            EventDataModel,
            id
        )
        await this.connection.manager.remove(event)
    }

    async getEvent(id: number) {
        return await this.connection.manager.findOne(EventDataModel, id)
    }

    async getEvents() {
        return await this.connection.manager.find(EventDataModel)
    }

    async getEventsFromClub(clubId: number) {
        return await this.connection.manager.find(EventDataModel, { clubId })
    }

    async updateEvent(input: {
        admissionFee?: number
        admissionFeeWithDiscount?: number
        amountOfFloors?: number
        clubId: number
        date: string
        description?: string
        genreIds?: number[]
        id: number
        imageUrl?: string
        link?: string
        minimumAge?: number
        name: string
        priceCategory?: number
        special?: string
    }) {
        const event = await this.connection.manager.findOne(
            EventDataModel,
            input.id
        )
        if (!event) return undefined

        event.admissionFee = input.admissionFee || null
        event.admissionFeeWithDiscount = input.admissionFeeWithDiscount || null
        event.amountOfFloors = input.amountOfFloors || null
        event.clubId = input.clubId
        event.date = input.date
        event.description = input.description || null
        event.imageUrl = input.imageUrl || null
        event.link = input.link || null
        event.minimumAge = input.minimumAge || null
        event.name = input.name
        event.priceCategory = input.priceCategory || null
        event.special = input.special || null

        await this.connection.manager.save(event)
    }
}
