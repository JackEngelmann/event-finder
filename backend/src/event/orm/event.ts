import {
    Connection,
    Entity,
    Column,
    PrimaryGeneratedColumn,
    EntityRepository,
    Repository,
} from 'typeorm'

@Entity('event')
export class EventDataModel {
    @Column({ type: 'decimal', nullable: true })
    admissionFee?: number | null

    @Column({ type: 'decimal', nullable: true })
    admissionFeeWithDiscount?: number | null

    @Column({ type: 'int', nullable: true })
    amountOfFloors?: number | null

    @Column()
    clubId!: number

    @Column()
    date!: string

    @Column({
        type: 'text',
        nullable: true,
        charset: 'utf8mb4',
        collation: 'utf8mb4_general_ci',
    })
    description?: string | null

    @PrimaryGeneratedColumn()
    id!: number

    // LEGACY
    @Column({ type: 'text', nullable: true })
    link?: string | null

    @Column({ type: 'int', nullable: true })
    minimumAge?: number | null

    @Column()
    name!: string

    @Column({ type: 'int', nullable: true })
    priceCategory?: number | null

    @Column({
        type: 'text',
        nullable: true,
        charset: 'utf8mb4',
        collation: 'utf8mb4_general_ci',
    })
    special?: string | null
}

@EntityRepository(EventDataModel)
export class EventRepository extends Repository<EventDataModel> {
    async createAndSave(input: {
        admissionFee?: number
        admissionFeeWithDiscount?: number
        amountOfFloors?: number
        clubId: number
        date: string
        description?: string
        genreIds?: number[]
        minimumAge?: number
        name: string
        priceCategory?: number
        special?: string
    }) {
        const event = this.create(input)
        return await this.save(event)
    }

    async findEventsForClub(clubId: number) {
        return await this.find({ clubId })
    }
}
