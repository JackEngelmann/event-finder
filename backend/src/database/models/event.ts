import { Database } from '../database'

export class EventDataModel {
    admissionFee?: number
    admissionFeeWithDiscount?: number
    amountOfFloors?: number
    clubId: string
    date: string
    description?: string
    id: number
    imageUrl?: string
    link?: string
    minimumAge?: number
    name: string
    priceCategory?: 1 | 2 | 3
    special?: string
    constructor(row: any) {
        this.admissionFee = row.admissionfee
        this.admissionFeeWithDiscount = row.admissionfeewithdiscount
        this.amountOfFloors = row.amountoffloors
        this.clubId = row.clubid
        this.date = row.date
        this.description = row.description
        this.id = row.id
        this.imageUrl = row.imageurl
        this.link = row.link
        this.minimumAge = row.minimumage
        this.name = row.name
        this.priceCategory = row.pricecategory
        this.special = row.special
    }
}

export class EventModel {
    private db: Database

    constructor(db: Database) {
        this.db = db
    }

    async createEvent(input: {
        admissionFee?: number
        admissionFeeWithDiscount?: number
        amountOfFloors?: number
        clubId: number
        date: string
        description?: String
        genreIds?: number[]
        imageUrl?: string
        link?: string
        minimumAge?: number
        name: string
        priceCategory?: number
        special?: string
    }) {
        return this.db
            .get(
                `
                INSERT INTO event (
                    admissionFee,
                    admissionFeeWithDiscount,
                    amountOfFloors,
                    clubId,
                    date,
                    description,
                    imageUrl,
                    link,
                    minimumAge,
                    name,
                    priceCategory,
                    special
                ) VALUES (
                    $1,
                    $2,
                    $3,
                    $4,
                    $5,
                    $6,
                    $7,
                    $8,
                    $9,
                    $10,
                    $11,
                    $12
                ) RETURNING Id
            `,
                [
                    input.admissionFee,
                    input.admissionFeeWithDiscount,
                    input.amountOfFloors,
                    input.clubId,
                    input.date,
                    input.description,
                    input.imageUrl,
                    input.link,
                    input.minimumAge,
                    input.name,
                    input.priceCategory,
                    input.special,
                ]
            )
            .then(res => res.id)
            .catch(err => {
                console.error(err)
            })
    }

    deleteEvent(id: number) {
        return this.db.run('DELETE FROM event WHERE id = $1', [id])
    }

    async getEvent(id: number) {
        const sql = 'SELECT * FROM event WHERE id = $1'
        const params = [id]
        const row = await this.db.get(sql, params)
        if (!row) return undefined
        const event = new EventDataModel(row)
        return event
    }

    async getEvents() {
        const sql = 'SELECT * FROM event'
        const rows = await this.db.all(sql, [])
        return rows.map(r => new EventDataModel(r))
    }

    async getEventsFromClub(clubId: number) {
        const sql = 'SELECT * FROM event WHERE clubId = $1'
        const rows = await this.db.all(sql, [clubId])
        return rows.map(r => new EventDataModel(r))
    }

    updateEvent(input: {
        admissionFee?: number
        admissionFeeWithDiscount?: number
        amountOfFloors?: number
        clubId: number
        date: string
        description?: String
        genreIds?: number[]
        id: number
        imageUrl?: string
        link?: string
        minimumAge?: number
        name: string
        priceCategory?: number
        special?: string
    }) {
        return this.db.run(
            `
                UPDATE event 
                SET
                    admissionFee = $1,
                    admissionFeeWithDiscount = $2,
                    amountOfFloors = $3,
                    clubId = $4,
                    date = $5,
                    description = $6,
                    imageUrl = $7,
                    link = $8,
                    minimumAge = $9,
                    name = $10,
                    priceCategory = $11,
                    special = $12
                WHERE id = $13
            `,
            [
                input.admissionFee,
                input.admissionFeeWithDiscount,
                input.amountOfFloors,
                input.clubId,
                input.date,
                input.description,
                input.imageUrl,
                input.link,
                input.minimumAge,
                input.name,
                input.priceCategory,
                input.special,
                input.id,
            ]
        )
    }
}
