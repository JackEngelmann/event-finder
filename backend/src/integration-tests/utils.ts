import { Database } from '../database/database'

export const genreFragment = `
    id
    name
`

export const eventFragment = `
    club {
        id
        name
    }
    name,
    description,
    date,
    id
    imageUrl,
    priceCategory,
    admissionFee,
    admissionFeeWithDiscount,
    link
    special,
    minimumAge,
    amountOfFloors
    genres {
        ${genreFragment}
    }
`

export const clubFragment = `
    address
    contact
    description
    email
    id
    link
    name
    region
    specials
`

export async function insertTestData(db: Database) {
    await db.run(`
        INSERT INTO club (
            name,
            address,
            region,
            contact,
            email,
            specials,
            description,
            link
        ) VALUES (
            'name',
            'address',
            'region',
            'contact',
            'email',
            'specials',
            'description',
            'link'
        )
    `)
    await db.run(`
        INSERT INTO club (
            name,
            address,
            region,
            contact,
            email,
            specials,
            description,
            link
        ) VALUES (
            'name',
            'address',
            'region',
            'contact',
            'email',
            'specials',
            'description',
            'link'
        )
    `)
    await db.run(`
        INSERT INTO event
        (
            name,
            description,
            date,
            clubId,
            imageUrl,
            priceCategory,
            admissionFee,
            admissionFeeWithDiscount,
            special,
            minimumAge,
            amountOfFloors
        ) 
        VALUES (
            'name',
            'description',
            'date',
            1,
            'imageUrl',
            2,
            10.0,
            20.0,
            'specials',
            18,
            4
        )
    `)
    await db.run(`
        INSERT INTO genre (name) VALUES
            ('genre1'),
            ('genre2'),
            ('genre3')
    `)
    return await db.run(`
        INSERT INTO eventGenre (eventId, genreId) VALUES (1, 1)
    `)
}
