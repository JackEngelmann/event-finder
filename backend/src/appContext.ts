import { Database } from "./database/database";
import { UserDataModel } from "./database/models/user";

export type AppContext = { db: Database, isAdmin: boolean, user?: UserDataModel }