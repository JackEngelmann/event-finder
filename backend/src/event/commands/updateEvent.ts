import { AppContext } from '../../infrastructure/appContext'
import { LinkType } from '../../link/orm/link'
import { setLinksForEvent } from '../../link/commands/setLinksForEvent'
import { EventRepository } from '../orm/event'
import { setGenresForEvent } from '../../genre/commands/setGenresForEvent'
import { setImageUrlsForEvent } from '../../image/commands/setImageUrlsForEvent'
import logger from '../../infrastructure/logger'

export type UpdateEventInput = {
    admissionFee?: number
    admissionFeeWithDiscount?: number
    amountOfFloors?: number
    clubId: number
    date: string
    description?: string
    genreIds?: number[]
    id: number
    imageUrls?: string[]
    links?: LinkType[]
    minimumAge?: number
    name: string
    priceCategory?: number
    special?: string
}

export async function updateEvent(appContext: AppContext, input: UpdateEventInput) {
    const { db } = appContext
    const eventRepository = db.getCustomRepository(EventRepository)
    try {
        await eventRepository.update(input.id, {
            admissionFee: input.admissionFee || null,
            admissionFeeWithDiscount:
                input.admissionFeeWithDiscount || null,
            amountOfFloors: input.amountOfFloors || null,
            clubId: input.clubId,
            date: input.date,
            description: input.description || null,
            id: input.id,
            minimumAge: input.minimumAge || null,
            name: input.name,
            priceCategory: input.priceCategory || null,
            special: input.special || null,
        })
        await setGenresForEvent(appContext, {
            eventId: input.id,
            genreIds: input.genreIds,
        })
        if (input.imageUrls) {
            await setImageUrlsForEvent(appContext, {
                eventId: input.id,
                imageUrls: input.imageUrls,
            })
        }
        if (input.links) {
            await setLinksForEvent(appContext, {
                eventId: input.id,
                links: input.links,
            })
        }
    } catch (error) {
        logger.error(error)
        throw error
    }
}
