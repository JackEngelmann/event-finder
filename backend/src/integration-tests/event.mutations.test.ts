import fetch from 'node-fetch'
import express from 'express'
import http from 'http'
import FormData from 'form-data'
import { ApolloServer } from 'apollo-server-express'
import { createTestDb, Database, destroyTestDb } from '../database/database'
import { insertTestData, eventFragment, clubFragment } from './utils'
import { createTestClient, ApolloServerTestClient } from 'apollo-server-testing'
import { typeDefs } from '../graphql/schema'
import { resolvers } from '../graphql/resolvers'
import { CreateEventInput } from '../mutations/createEvent'
import { UpdateEventInput } from '../mutations/updateEvent'
import { queryEvent } from '../queries/event'
import { AppContext } from '../appContext'
import { createReadStream } from 'fs'
import { join } from 'path'
import { GraphQLResponse } from 'apollo-server-core'

const DB_NAME = 'eventmutationdb'
const createEventMutation = `
    mutation CreateEvent($input: CreateEventInput!) {
        createEvent(input: $input) {
            event {
                ${eventFragment}
            }
        }
    }
`
const updateEventMutation = `
    mutation UpdateEvent($input: UpdateEventInput!) {
        updateEvent(input: $input) {
            event {
                ${eventFragment}
            }
        }
    }
`
const deleteEventMutation = `
    mutation DeleteEvent($id: Int!) {
        deleteEvent(id: $id) {
            id
        }
    }
`

describe('event mutations: ', () => {
    let server: ApolloServerTestClient | undefined = undefined
    let db: Database | undefined = undefined
    let appContext: AppContext = { db: db!, isAdmin: true }

    beforeEach(async done => {
        const newDb = await createTestDb(DB_NAME)
        db = newDb
        appContext = { db: newDb, isAdmin: true }
        await insertTestData(newDb)
        const newServer = new ApolloServer({
            typeDefs,
            resolvers,
            context: { db: newDb },
        })
        server = createTestClient(newServer)
        done()
    })

    afterEach(async done => {
        await destroyTestDb(DB_NAME)
        done()
    })

    describe('create event', () => {
        test('create an event with only required information', async done => {
            const input: CreateEventInput = {
                name: 'event-name',
                date: 'event-date',
                clubId: 1,
            }
            const result = await server!.mutate({
                mutation: createEventMutation,
                variables: { input },
            })
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
            done()
        })

        test('create an event with empty genres', async done => {
            const input: CreateEventInput = {
                name: 'event-name',
                date: 'event-date',
                clubId: 1,
                genreIds: [],
            }
            const result = await server!.mutate({
                mutation: createEventMutation,
                variables: { input },
            })
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
            done()
        })

        test('create an event with all fields', async done => {
            const input: CreateEventInput = {
                name: 'event-name',
                description: 'event-description',
                date: 'event-date',
                clubId: 1,
                genreIds: [1, 2],
                special: 'event-special',
                priceCategory: 2,
                admissionFee: 10.0,
                admissionFeeWithDiscount: 4.0,
                minimumAge: 18,
                link: 'http://www.some-link.de',
                amountOfFloors: 3,
            }
            const result = await server!.mutate({
                mutation: createEventMutation,
                variables: { input },
            })
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
            done()
        })
    })

    describe('update event', () => {
        test('to only have required fields', async done => {
            const input: UpdateEventInput = {
                id: 1,
                name: 'updated-name',
                date: 'updated-date',
                clubId: 2,
            }
            const result = await server!.mutate({
                mutation: updateEventMutation,
                variables: { input },
            })
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
            done()
        })

        test('return null for event when updating a non-existent event', async done => {
            const input: UpdateEventInput = {
                id: 10000,
                name: 'updated-name',
                date: 'updated-date',
                clubId: 2,
            }
            const result = await server!.mutate({
                mutation: updateEventMutation,
                variables: { input },
            })
            expect(result.data).toBeDefined()
            expect(result.data!.updateEvent).toBeDefined()
            expect(result.data!.updateEvent.event).toBeNull()
            done()
        })

        test('with empty genre list', async done => {
            const input: UpdateEventInput = {
                id: 1,
                name: 'updated-name',
                date: 'updated-date',
                clubId: 2,
                genreIds: [],
            }
            const result = await server!.mutate({
                mutation: updateEventMutation,
                variables: { input },
            })
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
            done()
        })

        test('with all fields', async done => {
            const input: UpdateEventInput = {
                id: 1,
                name: 'updated-name',
                date: 'updated-date',
                clubId: 2,
                description: 'updated-description',
                genreIds: [2, 3],
                special: 'updated-special',
                priceCategory: 1,
                admissionFee: 20.0,
                admissionFeeWithDiscount: 3.0,
                minimumAge: 16,
                amountOfFloors: 2,
                link: 'http://www.some-link.de',
            }
            const result = await server!.mutate({
                mutation: updateEventMutation,
                variables: { input },
            })
            expect(result.data).toBeDefined()
            expect(result).toMatchSnapshot()
            done()
        })
    })

    describe('delte event', () => {
        test('delete an event', async done => {
            const id = 1

            const event = await queryEvent(appContext, id)
            expect(event).toBeDefined()

            await server!.mutate({
                mutation: deleteEventMutation,
                variables: { id },
            })

            const deletedEvent = await queryEvent(appContext, id)
            expect(deletedEvent).toBeUndefined()
            done()
        })
    })
})

describe('event image field', () => {
    let server: ApolloServer
    let app: express.Application
    let httpServer: http.Server
    let port: number

    async function createServer() {
        const newDb = await createTestDb(DB_NAME)
        server = new ApolloServer({
            typeDefs,
            resolvers,
            context: { db: newDb },
        })
        app = express()
        server.applyMiddleware({ app })
        httpServer = await new Promise<http.Server>(resolve => {
            const l: http.Server = app.listen({ port: 0 }, () => resolve(l))
        })
        await insertTestData(newDb)
        port = (httpServer.address() as { port: number }).port
    }

    afterEach(async done => {
        if (server) await server.stop()
        if (httpServer) await httpServer.close()
        await destroyTestDb(DB_NAME)
        done()
    })

    test('update event with an image', async done => {
        await createServer()

        const input: any = {
            id: 1,
            name: 'updated-name',
            date: 'updated-date',
            clubId: 2,
            image: null,
        }

        const body = new FormData()
        body.append(
            'operations',
            JSON.stringify({
                query: updateEventMutation,
                variables: { input },
            })
        )
        body.append('map', JSON.stringify({ 1: ['variables.input.image'] }))
        body.append('1', createReadStream(join(__dirname, 'testPicture.jpg')))

        const result: GraphQLResponse = await fetch(`http://localhost:${port!}/graphql`, {
            method: 'POST',
            body: body as any,
        }).then(res => res.json())

        expect(result.errors).toBeUndefined()
        expect(result.data!.updateEvent.event.imageUrl).toBeDefined()
        done()
    })

    test('create event with an image', async done => {
        await createServer()

        const input: any = {
            name: 'created-name',
            date: 'created-date',
            clubId: 2,
            image: null,
        }

        const body = new FormData()
        body.append(
            'operations',
            JSON.stringify({
                query: createEventMutation,
                variables: { input },
            })
        )
        body.append('map', JSON.stringify({ 1: ['variables.input.image'] }))
        body.append('1', createReadStream(join(__dirname, 'testPicture.jpg')))

        const result: GraphQLResponse = await fetch(
            `http://localhost:${port!}/graphql`,
            {
                method: 'POST',
                body: body as any,
            }
        ).then(res => res.json())

        expect(result.errors).toBeUndefined()
        expect(result.data!.createEvent.event.imageUrl).toBeDefined()
        done()
    })
})
