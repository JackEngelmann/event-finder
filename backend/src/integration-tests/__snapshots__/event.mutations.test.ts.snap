// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event mutations:  create event create an event with all fields 1`] = `
Object {
  "data": Object {
    "createEvent": Object {
      "event": Object {
        "admissionFee": 10,
        "admissionFeeWithDiscount": 4,
        "amountOfFloors": 3,
        "club": Object {
          "id": 1,
          "name": "name",
        },
        "date": "event-date",
        "description": "event-description",
        "genres": Array [
          Object {
            "id": 1,
            "name": "genre1",
          },
          Object {
            "id": 2,
            "name": "genre2",
          },
        ],
        "id": 2,
        "imageUrl": null,
        "link": "http://www.some-link.de",
        "minimumAge": 18,
        "name": "event-name",
        "priceCategory": 2,
        "special": "event-special",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`event mutations:  create event create an event with empty genres 1`] = `
Object {
  "data": Object {
    "createEvent": Object {
      "event": Object {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": Object {
          "id": 1,
          "name": "name",
        },
        "date": "event-date",
        "description": null,
        "genres": Array [],
        "id": 2,
        "imageUrl": null,
        "link": null,
        "minimumAge": null,
        "name": "event-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`event mutations:  create event create an event with only required information 1`] = `
Object {
  "data": Object {
    "createEvent": Object {
      "event": Object {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": Object {
          "id": 1,
          "name": "name",
        },
        "date": "event-date",
        "description": null,
        "genres": Array [],
        "id": 2,
        "imageUrl": null,
        "link": null,
        "minimumAge": null,
        "name": "event-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`event mutations:  create event create an event without permission shoul yield error 1`] = `
Array [
  [GraphQLError: You need to be logged in as admin to mutate data],
]
`;

exports[`event mutations:  delte event without permission should yield error 1`] = `
Array [
  [GraphQLError: You need to be logged in as admin to mutate data],
]
`;

exports[`event mutations:  update event to only have required fields 1`] = `
Object {
  "data": Object {
    "updateEvent": Object {
      "event": Object {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": Object {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": null,
        "genres": Array [],
        "id": 1,
        "imageUrl": null,
        "link": null,
        "minimumAge": null,
        "name": "updated-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`event mutations:  update event with all fields 1`] = `
Object {
  "data": Object {
    "updateEvent": Object {
      "event": Object {
        "admissionFee": 20,
        "admissionFeeWithDiscount": 3,
        "amountOfFloors": 2,
        "club": Object {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": "updated-description",
        "genres": Array [
          Object {
            "id": 2,
            "name": "genre2",
          },
          Object {
            "id": 3,
            "name": "genre3",
          },
        ],
        "id": 1,
        "imageUrl": null,
        "link": "http://www.some-link.de",
        "minimumAge": 16,
        "name": "updated-name",
        "priceCategory": 1,
        "special": "updated-special",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`event mutations:  update event with emojis in description 1`] = `
Object {
  "data": Object {
    "updateEvent": Object {
      "event": Object {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": Object {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": "üçÉüîùüéâüì¢üëâüåµüéµ",
        "genres": Array [],
        "id": 1,
        "imageUrl": null,
        "link": null,
        "minimumAge": null,
        "name": "updated-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`event mutations:  update event with empty genre list 1`] = `
Object {
  "data": Object {
    "updateEvent": Object {
      "event": Object {
        "admissionFee": null,
        "admissionFeeWithDiscount": null,
        "amountOfFloors": null,
        "club": Object {
          "id": 2,
          "name": "name",
        },
        "date": "updated-date",
        "description": null,
        "genres": Array [],
        "id": 1,
        "imageUrl": null,
        "link": null,
        "minimumAge": null,
        "name": "updated-name",
        "priceCategory": null,
        "special": null,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`event mutations:  update event without permission should yield error 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You need to be logged in as admin to mutate data],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
